### PR TITLE
Database and Value Objects

### PR DESCRIPTION
**Pull request by bvisness**
Big refactoring of pretty much everything! Here are the highlights:
- Data is now saved into a Postgres database instead of Github Issues. They are stored as JSON for flexibility.
- All feedback items are now exchanged using the `FeedbackItem` class, which is build on [sparkphp/data](https://github.com/sparkphp/data) as an immutable value object.
- Message formatting is now in the `HipChatFormatter` class.

### CODE REVIEW

File: migrations/20151027181154_postgres_initial_migration.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+use Phinx\Migration\AbstractMigration;
+
+class PostgresInitialMigration extends AbstractMigration
```
Comment (shadowhand) [inline migrations/20151027181154_postgres_initial_migration.php:5]: touchedbygod

File: phinx.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+paths:
```
Comment (shadowhand) [inline phinx.yml]: Rename this to `phinx.yml.dist` so that it can be overloaded with a custom version.

File: src/Domain/FeedbackGetDomain.php [hunk 3]
```diff
@@ -29,22 +30,19 @@ public function __invoke(array $input)
                 $this->initRedis();
             }
 
-            $feedbackItems = $this->getFeedbackItems();
+            $rawFeedbacks = $this->getRawFeedbacks();
 
-            if (count($feedbackItems) > 0 && !$debug) {
-                $this->setRedisValue($this->getValueForRedis(reset($feedbackItems)));
+            if (count($rawFeedbacks) > 0 && !$debug) {
```
Comment (shadowhand) [inline src/Domain/FeedbackGetDomain.php]: You don't actually need a `count` comparison here. Using `if ($rawFeedbacks && ...)` would be enough.

File: src/Domain/FeedbackGetDomain.php [hunk 4]
```diff
@@ -29,22 +30,19 @@ public function __invoke(array $input)
...
+                $this->setRedisValue($this->getValueForRedis(reset($rawFeedbacks)));
```
Comment (shadowhand) [inline src/Domain/FeedbackGetDomain.php]: What's the `reset` for?

Comment (bvisness) [inline src/Domain/FeedbackGetDomain.php]: It gets the first element of the array.

File: src/Domain/FeedbackGetDomain.php (hunk 3 above)
Comment (bvisness) [inline src/Domain/FeedbackGetDomain.php]: Good catch, thanks.

File: src/FeedbackItem.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    private function types()
+    {
+        return [
+            'body' => 'string',
+            'source' => 'string',
+            'link' => 'string',
+            'title' => 'string',
+            'sender' => 'string',
+            'tone' => 'string'
```
Comment (shadowhand) [inline src/FeedbackItem.php:32]: Normally everything will be a string. Nothing wrong with being explicit, I just tend to skip typing anything that is a string.

File: src/FeedbackRating.php [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+namespace Wheniwork\Feedback;
+
+use Spark\Data\ArraySerializableInterface;
+
+class FeedbackRating implements ArraySerializableInterface
+{
+    private $rating;
+    private $max_rating;
+
+    public function __construct($rating, $max_rating) {
```
Comment (shadowhand) [inline src/FeedbackRating.php]: Missing newline before `{`

File: src/FeedbackRating.php [hunk 7]
```diff
@@ -0,0 +1,25 @@
+<?php
+namespace Wheniwork\Feedback;
+
+use Spark\Data\ArraySerializableInterface;
+
+class FeedbackRating implements ArraySerializableInterface
```
Comment (shadowhand) [inline src/FeedbackRating.php:7]: Any reason this isn't also an entity or at least an `ImmutableValueObject`?

Comment (bvisness) [inline src/FeedbackRating.php:7]: Probably not. I think I coded this up before I started using sparkphp/data.

File: src/Domain/FeedbackGetDomain.php (hunk 4 above)
Comment (shadowhand) [inline src/Domain/FeedbackGetDomain.php]: Ah, you could use `current` for that instead. It doesn't have side effects like `reset` does.

**Pull request merged**